weave/compile.py

```diff
@@ -477,10 +442,143 @@ def compile_quote(
     return graph.map_nodes_full(nodes, _quote_nodes_map_fn, on_error)
 
 
+def _needs_gql_propagation(node: graph.OutputNode) -> bool:
+    """Determines if a node needs to propagate the GQL keys of its inputs."""
+    fq_opname = node.from_op.full_name
+    opdef = registry_mem.memory_registry.get_op(fq_opname)
+    plugin = gql_op_plugin._get_gql_plugin(opdef)
+    first_arg_name = gql_op_plugin._first_arg_name(opdef)
+
+    if first_arg_name is None:
+        return False
+
+    first_arg_type: types.Type = node.from_op.inputs[first_arg_name].type
+    unwrapped_first_arg_type, _ = tagged_value_type_helpers.unwrap_tags(first_arg_type)
+
+    if opdef.derived_from and opdef.derived_from.derived_ops["mapped"] == opdef:
+        plugin = gql_op_plugin._get_gql_plugin(opdef.derived_from)
+        unwrapped_first_arg_type = typing.cast(
+            types.List, unwrapped_first_arg_type
+        ).object_type
+
+    return fq_opname.endswith("GQLResolver") or (
```

Comment:
It would be nice to not rely on this naming convention (since it is not enforced programatically.) Would it be possible to change this to something more structured?

Reply:
done